### PR TITLE
support intercept specified command handler

### LINKED ISSUE #2578: sentinel-transport-common HTTP security | sentinel-transport-common模块暴露的http接口没有访问验证，是否存在安全隐患问题
Java服务在外网使用中，sentinel-transport-common模块暴露的http接口没有访问验证，是否存在安全隐患问题，例如攻击者直接通过构造请求调用修改限流规则？

### PR DESCRIPTION
**Pull request by icodening**
### Describe what this PR does / why we need it
support intercept specified command handler
example, users can extend support for authentication, suitable for "Zero Trust" network architecture

### Does this pull request fix one issue?
Fixes #2578 

### Describe how you did it
wrap command handlers when based on conditions

### Describe how to verify it
implement 'com.alibaba.csp.sentinel.command.CommandHandlerInterceptor' interface, and add file 'com.alibaba.csp.sentinel.command.CommandHandlerInterceptor' in META-INF/services dir

### Special notes for reviews
example:
````java
public class GetRulesCommandHandlerInterceptor implements CommandHandlerInterceptor {

    @Override
    public boolean shouldIntercept(String commandName) {
        // intercept the handler whose @CommandMapping name is "getRules"
        return "getRules".equals(commandName);
    }

    @Override
    public CommandResponse intercept(CommandRequest commandRequest, CommandRequestExecution commandRequestExecution) {
        // before handle
        CommandResponse response = commandRequestExecution.execute(commandRequest);
        // after handle
        return response;
    }
}
````

### CODE REVIEW

**Comment by sczyh30** [comment]
Any suggestions on this PR? @cdfive @brotherlu-xcq @alibaba/sentinel-committer

**Review by brotherlu-xcq — CHANGES_REQUESTED** [review]
@icodening Nice work. but I think we need more discussion about your design. when one handler has more than one interceptors, I think we need execute all the interceptors if no one fail. and could you provide a flow chart to help us to know your idea?

**Comment by icodening** [comment]
> @icodening Nice work. but I think we need more discussion about your design. when one handler has more than one interceptors, I think we need execute all the interceptors if no one fail. and could you provide a flow chart to help us to know your idea?

this is a simple flow chart

image

**Comment by icodening** [comment]
> @icodening Nice work. but I think we need more discussion about your design. when one handler has more than one interceptors, I think we need execute all the interceptors if no one fail. and could you provide a flow chart to help us to know your idea?

this is a simple flow chart

image

File: sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java [hunk 1]
```diff
@@ -15,12 +15,13 @@
...
 package com.alibaba.csp.sentinel.command;
 
-import java.util.*;
-
 import com.alibaba.csp.sentinel.command.annotation.CommandMapping;
+import com.alibaba.csp.sentinel.command.handler.InterceptingCommandHandler;
 import com.alibaba.csp.sentinel.spi.SpiLoader;
 import com.alibaba.csp.sentinel.util.StringUtil;
 
+import java.util.*;
```
Comment (brotherlu-xcq) [inline sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java:23]: please remove the unnecessary changes.

File: sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java [hunk 2]
```diff
@@ -38,8 +39,18 @@
...
+            if (!commandHandlerInterceptors.isEmpty()) {
+                List<CommandHandlerInterceptor> interceptors = new ArrayList<>();
+                for (CommandHandlerInterceptor commandHandlerInterceptor : commandHandlerInterceptors) {
+                    if (commandHandlerInterceptor.shouldIntercept(name)) {
+                        interceptors.add(commandHandlerInterceptor);
+                    }
+                }
+               handler = new InterceptingCommandHandler(handler, interceptors);
+            }
             if (!StringUtil.isEmpty(name)) {
```
Comment (brotherlu-xcq) [inline sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java:54]: I think we can judge the name first. if name is empty, skip it quickly.

File: sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java [hunk 3]
```diff
@@ -38,8 +39,18 @@
...
         for (CommandHandler handler : handlers) {
             String name = parseCommandName(handler);
+            if (!commandHandlerInterceptors.isEmpty()) {
+                List<CommandHandlerInterceptor> interceptors = new ArrayList<>();
+                for (CommandHandlerInterceptor commandHandlerInterceptor : commandHandlerInterceptors) {
+                    if (commandHandlerInterceptor.shouldIntercept(name)) {
+                        interceptors.add(commandHandlerInterceptor);
+                    }
+                }
+               handler = new InterceptingCommandHandler(handler, interceptors);
```
Comment (brotherlu-xcq) [inline sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java:52]: if there is no interceptors mapping, I think we don't need the wrap the handler as `InterceptingCommandHandler`.

File: sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/handler/InterceptingCommandHandler.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * intercept specified command handler
+ *
+ * @author icodening
+ * @date 2022.03.03
+ */
+public class InterceptingCommandHandler<R> implements CommandHandler<R> {
```
Comment (brotherlu-xcq) [inline sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/handler/InterceptingCommandHandler.java:31]: Cloud we add an unit test and demo for the changes?

**Review by brotherlu-xcq — CHANGES_REQUESTED** [review]

File: sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java (hunk 1 above)
Comment (icodening) [reply sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java:23]: OK, I will slove it soon

File: sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java (hunk 3 above)
Comment (icodening) [reply sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java:52]: > if there is no interceptors mapping, I think we don't need the wrap the handler as `InterceptingCommandHandler`.

I don't understand here, I have called ````if (!commandHandlerInterceptors.isEmpty())```` before, that will not wrap as  ````InterceptingCommandHandler````

File: sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java (hunk 2 above)
Comment (icodening) [reply sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java:54]: > I think we can judge the name first. if name is empty, skip it quickly.

this is a good suggestion

File: sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java (hunk 3 above)
Comment (brotherlu-xcq) [reply sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java:52]: If no interceptors match, the handler is still wrapped as InterceptingCommandHandler.

Comment (icodening) [reply sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java:52]: > If no interceptors match, the handler is still wrapped as InterceptingCommandHandler.
OK, I understand this time

Comment (icodening) [reply sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/CommandHandlerProvider.java:52]: > If no interceptors match, the handler is still wrapped as InterceptingCommandHandler.

OK, I understand this time

File: sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/handler/InterceptingCommandHandler.java (hunk 4 above)
Comment (icodening) [reply sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/command/handler/InterceptingCommandHandler.java:31]: > Cloud we add an unit test and demo for the changes?

I am already add, please review

File: sentinel-demo/sentinel-demo-command-handler/pom.xml [hunk 5]
```diff
@@ -16,5 +16,9 @@
             <groupId>com.alibaba.csp</groupId>
             <artifactId>sentinel-transport-simple-http</artifactId>
         </dependency>
+        <dependency>
```
Comment (brotherlu-xcq) [inline sentinel-demo/sentinel-demo-command-handler/pom.xml:19]: the `sentinel-transport-simple-http` contains the  `sentinel-transport-common`. I think this import is unnecessary. do you think so?

**Review by brotherlu-xcq — CHANGES_REQUESTED** [review]

File: sentinel-demo/sentinel-demo-command-handler/pom.xml (hunk 5 above)
Comment (icodening) [reply sentinel-demo/sentinel-demo-command-handler/pom.xml:19]: > the `sentinel-transport-simple-http` contains the `sentinel-transport-common`. I think this import is unnecessary. do you think so?

I think implementation layer dependencies may use "provide" scope. Of cause, it can also be removed here

Comment (icodening) [reply sentinel-demo/sentinel-demo-command-handler/pom.xml]: I have removed, please review

**Review by brotherlu-xcq — APPROVED** [review]
LGTM

**Review by sczyh30 — APPROVED** [review]
LGTM

**Comment by sczyh30** [comment]
Thanks for contributing and look forward to more!
